lib/internal/modules/esm/loader.js

```diff
@@ -4,150 +4,176 @@
-    }
+  /**
+   * The index for assigning unique URLs to anonymous module evaluation
+   */
+  evalIndex = 0;
 
-    const { format } = getFormatResponse;
-    if (format === null) {
-      const dataUrl = RegExpPrototypeExec(
-        /^data:([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/,
-        url,
+  /**
+   * Registry of loaded modules, akin to `require.cache`
+   */
+  moduleMap = new ModuleMap();
+
+  /**
+   * Methods which translate input code or other information into es modules
+   */
+  translators = translators;
+
+  static pluckHooks({
+    globalPreload,
+    resolve,
+    load,
+    // obsolete hooks:
+    dynamicInstantiate,
+    getFormat,
+    getGlobalPreloadCode,
+    getSource,
+    transformSource,
+  }) {
+    const obsoleteHooks = [];
+    const acceptedHooks = ObjectCreate(null);
+
+    if (getGlobalPreloadCode) {
+      globalPreload ??= getGlobalPreloadCode;
+
+      process.emitWarning(
+        'Loader hook "getGlobalPreloadCode" has been renamed to "globalPreload"'
       );
-      throw new ERR_INVALID_MODULE_SPECIFIER(
-        url,
-        dataUrl ? `has an unsupported MIME type "${dataUrl[1]}"` : '');
-    }
-    if (typeof format !== 'string') {
-      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
-        'string', 'loader getFormat', 'format', format);
     }
+    if (dynamicInstantiate) obsoleteHooks.push('dynamicInstantiate');
+    if (getFormat) obsoleteHooks.push('getFormat');
+    if (getSource) obsoleteHooks.push('getSource');
+    if (transformSource) obsoleteHooks.push('transformSource');
 
-    if (format === 'builtin') {
-      return format;
-    }
+    if (obsoleteHooks.length) process.emitWarning(
+      `Obsolete loader hook(s) supplied: ${obsoleteHooks.join(', ')}`
+    );
 
-    if (this._resolve !== defaultResolve) {
-      try {
-        new URL(url);
-      } catch {
-        throw new ERR_INVALID_RETURN_PROPERTY(
-          'url', 'loader resolve', 'url', url
-        );
-      }
+    // Use .bind() to avoid giving access to the Loader instance when called.
+    if (globalPreload) {
+      acceptedHooks.globalPreloaders =
+        FunctionPrototypeBind(globalPreload, null);
+    }
+    if (resolve) {
+      acceptedHooks.resolver = FunctionPrototypeBind(resolve, null);
+    }
+    if (load) {
+      acceptedHooks.loader = FunctionPrototypeBind(load, null);
     }
 
-    if (this._resolve === defaultResolve &&
-      !StringPrototypeStartsWith(url, 'file:') &&
-      !StringPrototypeStartsWith(url, 'data:')
-    ) {
-      throw new ERR_INVALID_RETURN_PROPERTY(
-        'file: or data: url', 'loader resolve', 'url', url
+    return acceptedHooks;
+  }
+
+  /**
+   * Collect custom/user-defined hook(s). After all hooks have been collected,
+   * calls global preload hook(s).
+   * @param {object | object[]} customLoaders A list of exports from
+   * user-defined loaders (as returned by ESMLoader.import()).
+   */
+  async addCustomLoaders(
+    customLoaders = Array(0),
```

Comment:
FYI I did this as a V8 performance optimisation to avoid the known-unnecessary memory allocation. Please thumbs-down this comment if you still want the suggested change.